### PR TITLE
Nodejs: update core API routes

### LINKED ISSUE #1696: nodejs: update all API routes

### PR DESCRIPTION
**Pull request by Thoralf-M**
# Description of change

Please write a summary of your changes and why you made them. 

## Links to any relevant issues

Closes https://github.com/iotaledger/iota-sdk/issues/1696

## Type of change

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How the change has been tested

Running an example

### CODE REVIEW

File: bindings/nodejs/lib/client/client.ts [hunk 1]
```diff
@@ -463,15 +530,102 @@ export class Client {
...
+                index,
+            },
+        });
+        return JSON.parse(response).payload;
+    }
+
+    /**
+     * Get all UTXO changes of a given slot by commitment index.
+     *
+     * @param index Index of the commitment to look up.
```
Comment (DaughterOfMars) [inline bindings/nodejs/lib/client/client.ts]: perhaps we should call this `slot` instead

File: bindings/nodejs/lib/types/models/api/validators-response.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import type { HexEncodedString } from '../../utils/hex-encoding';
+
+/**
+ * Information of a validator.
+ */
+export interface ValidatorResponse {
+    /**
+     * Account address of the validator.
+     */
+    address: Bech32Address;
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/models/api/validators-response.ts:15]: I need you to remind me when/why to use `!` or not again. Why are you using it everywhere in `UtxoChangesResponse` for example and nowhere here?

File: bindings/nodejs/lib/types/models/state.ts [hunk 3]
```diff
@@ -13,9 +13,15 @@ export declare type BlockState =
 
 /**
  * The different states of a transaction.
+ * If 'pending', the transaction is not included yet.
+ * If 'accepted', the transaction is included.
+ * If 'confirmed' means transaction is included and its included block is confirmed.
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/models/state.ts]:
```suggestion
 * If 'confirmed', the transaction is included and its included block is confirmed.
```
For consistency, same for the 2 others

File: bindings/nodejs/lib/types/models/api/validators-response.ts (hunk 2 above)
Comment (Thoralf-M) [reply bindings/nodejs/lib/types/models/api/validators-response.ts:15]: One can only use it in classes to show the compiler that the value can't be undefined or null
UtxoChangesResponse could also be an interface I guess, not sure what's better here

File: sdk/src/types/api/core.rs [hunk 4]
```diff
@@ -356,11 +356,13 @@ pub enum BlockState {
...
-    // Stored but not confirmed.
+    // Not included yet.
     Pending,
-    // Confirmed with the first level of knowledge.
+    // Included.
+    Accepted,
+    // Included and its included block is confirmed.
     Confirmed,
-    // Included and can no longer be reverted.
+    // Included, its included block is finalized and cannot be reverted anymore.
```
Comment (Alex6323) [inline sdk/src/types/api/core.rs:365]:
```suggestion
    // Included and its confirmed block is finalized (cannot be reverted anymore).
```
at this point the block is confirmed, right?

File: sdk/src/wallet/operations/syncing/transactions.rs [hunk 5]
```diff
@@ -103,8 +103,10 @@ where
                     Ok(metadata) => {
                         if let Some(tx_state) = metadata.transaction_metadata.map(|m| m.transaction_state) {
                             match tx_state {
-                                // TODO: Separate TransactionState::Finalized?
-                                TransactionState::Finalized | TransactionState::Confirmed => {
+                                // TODO: Separate TransactionState::Finalized, TransactionState::Accepted?
```
Comment (Alex6323) [inline sdk/src/wallet/operations/syncing/transactions.rs]: is there an issue already to answer this question?

File: sdk/src/wallet/operations/syncing/transactions.rs [hunk 6]
```diff
@@ -103,8 +103,10 @@ where
...
+                                TransactionState::Finalized
+                                | TransactionState::Accepted
+                                | TransactionState::Confirmed => {
```
Comment (Alex6323) [inline sdk/src/wallet/operations/syncing/transactions.rs]: don't care that much, but consider keeping them in order: Accepted, Confirmed, Finalized?

File: bindings/core/src/method/client.rs [hunk 7]
```diff
@@ -282,17 +285,17 @@ pub enum ClientMethod {
     /// Look up a commitment by a given commitment index.
     GetCommitmentByIndex {
         /// Index of the commitment to look up.
-        index: SlotIndex,
+        slot: SlotIndex,
```
Comment (Alex6323) [inline bindings/core/src/method/client.rs:288]: TIP not updated yet? It still says `index`.

Comment (Thoralf-M) [reply bindings/core/src/method/client.rs:288]: No, commented already https://github.com/iotaledger/tips/pull/153/files#r1445850201

File: sdk/src/types/api/core.rs (hunk 4 above)
Comment (Thoralf-M) [reply sdk/src/types/api/core.rs:365]: The block is finalized :trollface: Don't think confirmed is right there, I'd say included is also not very clear, but I just copied like this from the TIP

File: bindings/python/iota_sdk/types/block/metadata.py [hunk 8]
```diff
@@ -49,15 +49,17 @@ class TransactionState(Enum):
...
 
     Attributes:
-        Pending: Stored but not confirmed.
-        Confirmed: Confirmed with the first level of knowledge.
-        Finalized: Included and can no longer be reverted.
-        Failed: The block is not successfully issued due to failure reason.
+        Pending: Not included yet.
+        Accepted: Included.
+        Confirmed: Included and its included block is confirmed.
+        Finalized: Included, its included block is finalized and cannot be reverted anymore.
```
Comment (Alex6323) [inline bindings/python/iota_sdk/types/block/metadata.py:55]: same change suggestion as on Rust side

File: sdk/src/wallet/operations/syncing/transactions.rs (hunk 6 above)
Comment (Thoralf-M) [reply sdk/src/wallet/operations/syncing/transactions.rs]: > don't care that much


Enough to comment :trollface: Will change

File: bindings/nodejs/lib/client/client.ts [hunk 9]
```diff
@@ -106,6 +117,74 @@ export class Client {
         return JSON.parse(response).payload;
     }
 
+    /**
+     * Check the readiness of the node to issue a new block, the reference mana cost based on the rate setter and
+     * current network congestion, and the block issuance credits of the requested account.
+     */
+    async getAccountCongestion(
+        accountId: AccountId,
```
Comment (Alex6323) [inline bindings/nodejs/lib/client/client.ts:125]: Ideally the function parameter names would be the same as in the TIP, right?

File: bindings/nodejs/lib/types/client/bridge/index.ts [hunk 10]
```diff
@@ -22,11 +26,17 @@ import type {
...
     __GetBlockRawMethod__,
     __GetIncludedBlockMethod__,
     __GetIncludedBlockMetadataMethod__,
+    __GetTransactionMetadataMethod__,
+    __GetCommitmentMethod__,
+    __GetUtxoChangesMethod__,
+    __GetUtxoChangesFullMethod__,
+    __GetCommitmentByIndexMethod__,
+    __GetUtxoChangesByIndexMethod__,
+    __GetUtxoChangesFullByIndexMethod__,
```
Comment (Alex6323) [inline bindings/nodejs/lib/types/client/bridge/index.ts:39]: Is there any order here that we adhere to?

File: sdk/src/wallet/operations/syncing/transactions.rs (hunk 5 above)
Comment (Thoralf-M) [reply sdk/src/wallet/operations/syncing/transactions.rs]: Added one

File: bindings/nodejs/lib/types/client/bridge/index.ts (hunk 10 above)
Comment (Thoralf-M) [reply bindings/nodejs/lib/types/client/bridge/index.ts:39]: Not really, but I try to do it in the same order as in the client methods

File: bindings/nodejs/lib/client/client.ts (hunk 9 above)
Comment (Thoralf-M) [reply bindings/nodejs/lib/client/client.ts:125]: If it makes more sense, but bech32 address could be also Ed25519 or something, which wouldn't work. That's why we have the accountId here

File: bindings/nodejs/lib/types/models/block-metadata.ts [hunk 11]
```diff
@@ -19,18 +20,14 @@ export interface IBlockMetadata {
...
-     * The transaction state.
-     */
-    transactionState?: TransactionState;
     /**
      * The block failure reason.
      */
     blockFailureReason?: BlockFailureReason;
     /**
-     * The transaction failure reason.
+     * The metadata of the transactions in the block.
```
Comment (Alex6323) [inline bindings/nodejs/lib/types/models/block-metadata.ts]:
```suggestion
     * The metadata of the transaction in the block.
```

File: sdk/src/types/api/core.rs (hunk 4 above)
Comment (Alex6323) [reply sdk/src/types/api/core.rs:365]: ah, I think you're right ... finalized can also mean definitely not confirmed and never going to be, right?

File: bindings/python/iota_sdk/types/block/metadata.py (hunk 8 above)
Comment (Alex6323) [reply bindings/python/iota_sdk/types/block/metadata.py:55]: sorry, no

File: bindings/nodejs/lib/client/client.ts (hunk 9 above)
Comment (Alex6323) [reply bindings/nodejs/lib/client/client.ts:125]: yes, and I like that much more than in the TIP

File: sdk/src/types/api/core.rs (hunk 4 above)
Comment (Thoralf-M) [reply sdk/src/types/api/core.rs:365]: It's first confirmed and then it gets finalized and when it's finalized the state can never change anymore

File: bindings/nodejs/lib/types/models/state.ts [hunk 12]
```diff
@@ -13,9 +13,15 @@ export declare type BlockState =
 
 /**
  * The different states of a transaction.
+ * If 'pending', the transaction is not included yet.
+ * If 'accepted', the transaction is included.
+ * If 'confirmed' the transaction is included and its included block is confirmed.
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/models/state.ts]:
```suggestion
 * If 'confirmed', the transaction is included and its included block is confirmed.
```

File: bindings/nodejs/lib/types/models/state.ts [hunk 13]
```diff
@@ -13,9 +13,15 @@ export declare type BlockState =
...
+ * If 'finalized' the transaction is included, its included block is finalized and cannot be reverted anymore.
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/models/state.ts]:
```suggestion
 * If 'finalized', the transaction is included, its included block is finalized and cannot be reverted anymore.
```

File: bindings/nodejs/lib/types/models/state.ts [hunk 14]
```diff
@@ -13,9 +13,15 @@ export declare type BlockState =
...
+ * If 'failed' the transaction is not successfully issued due to failure reason.
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/models/state.ts]:
```suggestion
 * If 'failed', the transaction is not successfully issued due to failure reason.
```

File: bindings/nodejs/lib/types/models/api/validators-response.ts (hunk 2 above)
Comment (thibault-martinez) [reply bindings/nodejs/lib/types/models/api/validators-response.ts:15]: So what's the rationale with `ValidatorResponse` being an interface and `UtxoChangesFullResponse` being a class for example? Are they used differently?

File: bindings/nodejs/lib/types/models/api/congestion-response.ts [hunk 15]
```diff
@@ -0,0 +1,27 @@
+// Copyright 2023 IOTA Stiftung
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/models/api/congestion-response.ts]: Maybe you want to update to 2024, just resolve if you don't

File: bindings/nodejs/lib/types/models/api/validators-response.ts (hunk 2 above)
Comment (Thoralf-M) [reply bindings/nodejs/lib/types/models/api/validators-response.ts:15]: `UtxoChangesFullResponse` contains outputs for which we want a class so one can call the output methods on it, the fields in `ValidatorResponse` you just access as they are, but you don't call any methods on them I would say

Comment (thibault-martinez) [reply bindings/nodejs/lib/types/models/api/validators-response.ts:15]: Ok I see, thanks

File: bindings/nodejs/lib/types/models/api/mana-rewards-response.ts [hunk 16]
```diff
@@ -0,0 +1,31 @@
+// Copyright 2023 IOTA Stiftung
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/models/api/mana-rewards-response.ts]:
```suggestion
// Copyright 2024 IOTA Stiftung
```

File: bindings/nodejs/lib/types/models/api/utxo-changes-response.ts [hunk 17]
```diff
@@ -0,0 +1,60 @@
+// Copyright 2023 IOTA Stiftung
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/models/api/utxo-changes-response.ts]:
```suggestion
// Copyright 2024 IOTA Stiftung
```

File: bindings/nodejs/lib/types/models/api/validators-response.ts [hunk 18]
```diff
@@ -0,0 +1,64 @@
+// Copyright 2023 IOTA Stiftung
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/models/api/validators-response.ts]:
```suggestion
// Copyright 2024 IOTA Stiftung
```

File: bindings/nodejs/lib/types/models/block-metadata.ts [hunk 19]
```diff
@@ -46,3 +43,21 @@ export interface IBlockWithMetadata {
      */
     metadata: IBlockMetadata;
 }
+
+/**
+ * Returns the metadata of a given transaction.
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/models/block-metadata.ts]: Bit weird to says returns on a type definition

**Review by DaughterOfMars — DISMISSED** [review]

File: bindings/nodejs/lib/types/models/api/mana-rewards-response.ts [hunk 20]
```diff
@@ -0,0 +1,31 @@
+// Copyright 2024 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+import { EpochIndex } from '../../block/slot';
+import { u64 } from '../../utils';
+
+/**
+ * Returns the mana rewards of an account or delegation output.
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/models/api/mana-rewards-response.ts]: There is more of these `Returns` comments

File: bindings/nodejs/lib/types/models/api/validators-response.ts [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+    latestSupportedProtocolHash: HexEncodedString;
+}
+
+/**
+ * A paginated list of all registered validators ready for the next epoch and indicates if they were active recently
+ * (are eligible for committee selection).
+ */
+export interface ValidatorsResponse {
+    /**
+     * ist of registered validators ready for the next epoch.
```
Comment (thibault-martinez) [inline bindings/nodejs/lib/types/models/api/validators-response.ts]: :trollface:

File: bindings/nodejs/lib/types/models/api/mana-rewards-response.ts (hunk 20 above)
Comment (Thoralf-M) [reply bindings/nodejs/lib/types/models/api/mana-rewards-response.ts]: All from the TIP :runner:

Comment (Thoralf-M) [reply bindings/nodejs/lib/types/models/api/mana-rewards-response.ts]: Updated

**Review by thibault-martinez — APPROVED** [review]

**Review by DaughterOfMars — APPROVED** [review]

File: bindings/nodejs/lib/client/client.ts (hunk 9 above)
Comment (Thoralf-M) [reply bindings/nodejs/lib/client/client.ts:125]: Can this be closed?

**Review by Alex6323 — APPROVED** [review]

**Pull request merged**